src/neuroglancer/sliceview/compresso/compresso.hpp

```diff
@@ -0,0 +1,593 @@
+
+					WINDOW value = window_values[windows[block]];
+					boundaries[iv] = (value >> offset) & 0b1;
+				}				
+			}
+			else {
+				for (size_t x = 0; x < sx; x++) {
+					size_t iv = x + sx * y + sxy * z;
+					xblock = x / xstep;
+					xoffset = x % xstep;
+					
+					size_t block = xblock + yblock + zblock;
+					size_t offset = xoffset + yoffset + zoffset;
+
+					WINDOW value = window_values[windows[block]];
+					boundaries[iv] = (value >> offset) & 0b1;
+				}
+			}
+		}
+	}
+
+	return boundaries;
+}
+
+template <typename LABEL>
+void decode_nonboundary_labels(
+		uint32_t *components, const std::vector<LABEL> &ids, 
+		const size_t sx, const size_t sy, const size_t sz,
+		LABEL* output
+) {
+	const size_t voxels = sx * sy * sz;
+	for (size_t i = 0; i < voxels; i++) {
+		output[i] = ids[components[i]];
+	}
+}
+
+template <typename LABEL>
+int decode_indeterminate_locations(
+		bool *boundaries, LABEL *labels, 
+		const std::vector<LABEL> &locations, 
+		const size_t sx, const size_t sy, const size_t sz,
+		const size_t connectivity
+) {
+	const size_t sxy = sx * sy;
+
+	size_t loc = 0;
+	size_t index = 0;
+
+	// go through all coordinates
+	for (size_t z = 0; z < sz; z++) {
+		for (size_t y = 0; y < sy; y++) {
+			for (size_t x = 0; x < sx; x++) {
+				loc = x + sx * y + sxy * z;
+
+				if (!boundaries[loc]) {
+					continue;
+				}
+				else if (x > 0 && !boundaries[loc - 1]) {
+					labels[loc] = labels[loc - 1];
+					continue;
+				}
+				else if (y > 0 && !boundaries[loc - sx]) {
+					labels[loc] = labels[loc - sx];
+					continue;
+				}
+				else if (connectivity == 6 && z > 0 && !boundaries[loc - sxy]) {
+					labels[loc] = labels[loc - sxy];
+					continue;
+				}
+				else if (locations.size() == 0) {
+					return 1;
+				}
+				
+				size_t offset = locations[index];
+
+				if (offset == 0) {
+					if (x == 0) {
+						return 2;
+					}
+					labels[loc] = labels[loc - 1];
+				}
+				else if (offset == 1) {
+					if (x >= sx - 1) {
+						return 3;
+					}
+					labels[loc] = labels[loc + 1];
+				}
+				else if (offset == 2) {
+					if (y == 0) {
+						return 4;
+					}
+					labels[loc] = labels[loc - sx];
+				}
+				else if (offset == 3) {
+					if (y >= sy - 1) {
+						return 5;
+					}
+					labels[loc] = labels[loc + sx];
+				}
+				else if (offset == 4) {
+					if (z == 0) {
+						return 6;
+					}
+					labels[loc] = labels[loc - sxy];
+				}
+				else if (offset == 5) {
+					if (z >= sz - 1) {
+						return 7;
+					}
+					labels[loc] = labels[loc + sxy];
+				}
+				else if (offset == 6) {
+					labels[loc] = locations[index + 1];
+					index++;
+				}
+				else {
+					labels[loc] = offset - 7;
+				}
+				index++;
+			}
+		}
+	}
+
+	return 0;
+}
+
+template <typename LABEL, typename WINDOW>
+int decompress(unsigned char* buffer, size_t num_bytes, LABEL* output = NULL) {
+
+	if (num_bytes < CompressoHeader::header_size) {
+		return 8;
+	}
+	else if (!CompressoHeader::valid_header(buffer)) {
```

Comment:
Could pass in const CompressorHeader &header to this function, rather than re-validating and re-parsing it.

Reply:
I could, but the parsing overhead is very low (probably < 1msec) and I didn't want to mess up the function sig.